### PR TITLE
working changes to NN library- BUG of file not found exception I am w…

### PR DESCRIPTION
**Pull request by chanacham**
Working on current bug of a file not found exception

### CODE REVIEW

File: src/test/java/smartsnake/neuralnetwork/chambre/NeuralNetworkBrainTest.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import smartsnake.Snake;
+
+import java.awt.*;
+import java.io.IOException;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.mock;
+
+class NeuralNetworkBrainTest {
```
Comment (andrewoid) [inline src/test/java/smartsnake/neuralnetwork/chambre/NeuralNetworkBrainTest.java:16]: Don't write a test for this

File: src/main/java/smartsnake/neuralnetwork/chambre/LearnAndSave.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        for (int i = 0; i < 50_000; i++) {
+            AstarBrain astarBrain = new AstarBrain();
+            Garden garden = new Garden(20, 15);
+            Snake snake = garden.getSnake();
+            snake.setBrain(astarBrain);
+
+            while (true) {
+
+                boolean gardenTick = garden.tick();
+                double[] gardenInput = neuralNetworkDataFactory.toInput(garden);
```
Comment (andrewoid) [inline src/main/java/smartsnake/neuralnetwork/chambre/LearnAndSave.java:24]: This line needs to be before tick()

File: src/main/java/smartsnake/neuralnetwork/chambre/LearnAndSave.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        NeuralNetwork neuralNetwork = new NeuralNetwork(5, 256, 4);
+        NeuralNetworkDataFactory neuralNetworkDataFactory = new NeuralNetworkDataFactory();
+
+        for (int i = 0; i < 50_000; i++) {
+            AstarBrain astarBrain = new AstarBrain();
+            Garden garden = new Garden(20, 15);
+            Snake snake = garden.getSnake();
+            snake.setBrain(astarBrain);
+
+            while (true) {
```
Comment (andrewoid) [inline src/main/java/smartsnake/neuralnetwork/chambre/LearnAndSave.java:21]: `while (gardenTick)`

**Comment by chanacham** [comment]
After making these changes I attempted to run the Learn and Save but its going on 1.5 hours and I am not sure to stop it or this is what it is supposed to be

**Comment by chanacham** [comment]
NN Run Config

**Comment by chanacham** [comment]
Uploading SmartSnake – LearnAndSave.java [SmartSnake.main] 2023-12-13 15-38-44.mp4…

File: src/main/java/smartsnake/neuralnetwork/chambre/LearnAndSave.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        NeuralNetwork neuralNetwork = new NeuralNetwork(5, 256, 4);
+        NeuralNetworkDataFactory neuralNetworkDataFactory = new NeuralNetworkDataFactory();
+
+        for (int i = 0; i < 50_000; i++) {
+            AstarBrain astarBrain = new AstarBrain();
+            Garden garden = new Garden(20, 15);
+            Snake snake = garden.getSnake();
+            snake.setBrain(astarBrain);
+
+            while (garden.tick()) {
```
Comment (andrewoid) [inline src/main/java/smartsnake/neuralnetwork/chambre/LearnAndSave.java:21]: You are calling `garden.tick()` twice. Don't do that.

File: src/main/java/smartsnake/neuralnetwork/chambre/LearnAndSave.java [hunk 5]
```diff
@@ -18,9 +18,10 @@ public static void main(String[] args) {
             Snake snake = garden.getSnake();
             snake.setBrain(astarBrain);
 
-            while (garden.tick()) {
+            boolean gardenTick = garden.tick();
```
Comment (andrewoid) [inline src/main/java/smartsnake/neuralnetwork/chambre/LearnAndSave.java:21]: `boolean gardenTick = true;`
Only call garden.tick() inside the loop.

File: src/main/java/smartsnake/neuralnetwork/chambre/LearnAndSave.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+            boolean gardenTick = garden.tick();
+            while (gardenTick) {
+                double[] gardenInput = neuralNetworkDataFactory.toInput(garden);
+                gardenTick = garden.tick();
+                double[] snakeOutput = neuralNetworkDataFactory.toOutput(snake.getDirection());
+                neuralNetwork.train(gardenInput, snakeOutput);
+                if (!gardenTick) {
+                    break;
+                }
```
Comment (andrewoid) [inline src/main/java/smartsnake/neuralnetwork/chambre/LearnAndSave.java:29]: Once you make the above change then this is not needed anymore.

**Review by andrewoid — APPROVED** [review]

**Pull request merged**